src/main/java/org/openlmis/integration/dhis2/service/indicator/IndicatorService.java

```diff
@@ -0,0 +1,113 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
+ */
+
+package org.openlmis.integration.dhis2.service.indicator;
+
+import java.math.BigDecimal;
+import java.time.ZonedDateTime;
+import org.openlmis.integration.dhis2.domain.enumerator.IndicatorEnum;
+import org.openlmis.integration.dhis2.exception.ValidationMessageException;
+import org.openlmis.integration.dhis2.i18n.MessageKeys;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.util.Pair;
+import org.springframework.stereotype.Service;
+
+@Service
+public class IndicatorService {
+
+  @Autowired
+  OpeningBalance openingBalance;
+
+  @Autowired
+  ClosingBalance closingBalance;
+
+  @Autowired
+  ReceivedBalance receivedBalance;
+
+  @Autowired
+  PositiveAdjustment positiveAdjustment;
+
+  @Autowired
+  NegativeAdjustment negativeAdjustment;
+
+  /**
+   * Counts quantity of items for a given indicator enumerator.
+   *
+   * @param indicatorEnum Enumerator used to calculate indicators
+   * @param period Pair containing starting date and end date
+   * @return BigDecimal with generated indicator value for a given period
+   */
+  public BigDecimal generate(String source, IndicatorEnum indicatorEnum, Pair<ZonedDateTime,
+          ZonedDateTime> period, String orderable, String facility) {
+    String calculatedIndicator = "";
+    switch (indicatorEnum) {
+      case OPENING_BALANCE:
+        calculatedIndicator = openingBalance.calculateValue(
+                source, period, orderable, facility);
+        break;
+      case RECEIVED:
+        calculatedIndicator = receivedBalance.calculateValue(
+                source, period, orderable, facility);
+        break;
+      case CLOSING_BALANCE:
+        calculatedIndicator = closingBalance.calculateValue(
+                source, period, orderable, facility);
+        break;
+      case CCE_ALLOCATED:
+        break;
+      case CCE_OPERATIONAL:
+        break;
+      case NEGATIVE_ADJUSTMENTS:
+        calculatedIndicator = negativeAdjustment.calculateValue(
+                source, period, orderable, facility);
+        break;
+      case POSITIVE_ADJUSTMENTS:
+        calculatedIndicator = positiveAdjustment.calculateValue(
+                source, period, orderable, facility);
+        break;
+      case ADJUSTMENTS_BY_REASON:
+        break;
+      default:
+        throw new ValidationMessageException(MessageKeys.ERROR_ENUMERATOR_NOT_EXIST);
+    }
+
+    return new BigDecimal(calculatedIndicator);
```

Comment:
There is that thing which should be considered when you work with BigDecimals - a MathContext. 
It describes how to round values and how many 'significant' numbers there are (precision). It helps to define MathContext explicitly, maybe one context is enough for application, maybe you can decide to use one of the built-ins like MathContext.DECIMAL64. I think currently, when you would try to use the result of this method and do some maths, you'll get exception when rounding would be needed.

I would just move the creation of BigDecimal to the #calculateValue methods, this solution with String is strange, might be confusing.

Reply:
I moved BigDecimal creation to calculateValue methods and added MathContext to all of them